Extend the documentation with new custom tags.

Elv13 opened the pull request:
This is to allow @Aire-One to be unblocked with the inheritance support. There is now a new `@baseclass` tag we can define in the `@DOC_COMMON@` entries. It can then be accessed using `item.inherited` and `item.baseclass` in the template.

I also add a generic parser so we can add as many custom tags as we want. It re-implement some code which cannot be accessed from the template or the config. So now parts of ldoc are re-implemented in the ldoc config... However it adds some flexibility. An extended version of this will be able to auto-generate code examples for signal connections and other goodies by adding more custom attributes. It also allows custom tags to be rendered in tables rather than trees:

![Screenshot_20191124_040455](https://user-images.githubusercontent.com/340384/69492309-226dc780-0e6f-11ea-9bd4-3af04478b629.png)

@Aire-One If you think this is the way to go, I can merge this and then I will go ahead and update the entire API to use these (that will be painful...). It will slightly conflict with your PR.

I will also add some basic JavaScirpt to hide those new sections by default behind a "show more" summary. If I was to add this to everything and leave them visible, the signal to noise ratio would worsen.

edit:

```lua
--- The padding between the outline and the progressbar.
--@DOC_wibox_container_arcchart_paddings_EXAMPLE@
-- @property paddings
-- @tparam[opt=0] table|number paddings A number or a table
-- @tparam[opt=0] number paddings.top
-- @tparam[opt=0] number paddings.bottom
-- @tparam[opt=0] number paddings.left
-- @tparam[opt=0] number paddings.right
-- @emits [opt="optional_value"] property::paddings When the `paddings` changes.
-- @emitstparam property::paddings widget self The object being modified.
-- @emitstparam property::paddings table paddings The new paddings.
-- @usebeautiful beautiful.arcchart_paddings Fallback value when the object `paddings` isn't specified.

```

Aire-One commented on docs/config.ld:
```diff
@@ -93,6 +93,182 @@ new_type("filterfunction", "List filters", false)
+-- @emits class signal Message[...]
+add_custom_tag {
+    name   = "emits",
+    title  = "Emit signals:",
```
The `:` is already added by the template. (BTW I think they are mostly useless IMO considering they are `<h3>` so they seen as titles).

```suggestion
    title  = "Emit signals",
```

Elv13 replied:
done

Aire-One commented on docs/config.ld:
```diff
@@ -93,6 +93,182 @@ new_type("filterfunction", "List filters", false)
+        end
+
+        if params.type then
+            ret = ret .. '<span class="parameter">( '..
```
```suggestion
            ret = ret .. '<span class="parameter">('..
```

Elv13 replied:
done

Aire-One commented on docs/config.ld:
```diff
@@ -93,6 +93,182 @@ new_type("filterfunction", "List filters", false)
+        if params.type then
+            ret = ret .. '<span class="parameter">( '..
+                params.type.value..
+            " )</span>"
```
```suggestion
            ")</span>"
```

Elv13 replied:
done

Aire-One commented on docs/config.ld:
```diff
@@ -93,6 +93,182 @@ new_type("filterfunction", "List filters", false)
+
+        if params.type then
+            ret = ret .. '<span class="parameter">( '..
+                params.type.value..
```
```suggestion
                md("`"..params.type.value.."`")
```

Elv13 replied:
done

Aire-One commented:
I PR-ed your PR to add the missing part to match my inheritance documentation goal on #2921. 

Thanks again for your work!

> then I will go ahead and update the entire API to use these (that will be painful...).

I'll use these new tags while updating the widgets documentation in my PR. So you can remove them from your workload.

Elv13 commented:
Rebased with your suggested changes and some of my local commits. I added 2 new tags which do exactly as the `@emits` and `@usebeautiful`, but auto-generate the entire boilerplate for standard properties. Given I am pushing for those standards in other PRs, it makes sense to leverage them here. The code is pretty meta, so it doesn't add too much (code) weight beside making the build noticeably slower.

I didn't add the commits which add the "tag cloud" next to the item name. I am still evaluating different designs (and I suck at modern CSS, my last webdev work was when PHP4 was the cool kid favorite tech).

Thanks for working on this. I guess the best way forward is to merge this very soon so we stop diverging and iterate toward a better template. Given the inheritance stuff got auto merged (and I am ok with that, it's easier to iterate when things get merged), we are already down that path. So unless there is an outcry about how ugly these hacks are, I will merge this tomorrow.

actionless commented on docs/config.ld:
```diff
@@ -93,6 +93,257 @@ new_type("filterfunction", "List filters", false)
+    end
+}
+
+-- List the beautiful variables used by the method or property fllbacks.
```
```suggestion
-- List the beautiful variables used by the method or property fallbacks.
```

Elv13 commented:
Merging since the only commit that fail are due to the flacky geometry tests.
